src/Components/WebAssembly/Server/src/ComponentsWebAssemblyEndpointRouteBuilderExtensions.cs

```diff
@@ -107,5 +128,71 @@ private static void AddMapping(FileExtensionContentTypeProvider provider, string
                 provider.Mappings.Add(name, mimeType);
             }
         }
+
+        private static void NegotiateEncoding(HttpContext context, IWebHostEnvironment webHost)
+        {
+            var accept = context.Request.Headers[HeaderNames.AcceptEncoding];
+            if (StringValues.IsNullOrEmpty(accept))
+            {
+                return;
+            }
+
+            if (!StringWithQualityHeaderValue.TryParseList(accept, out var encodings) || encodings.Count == 0)
+            {
+                return;
+            }
+
+            var supportedEncodings = new HashSet<StringSegment>(StringSegmentComparer.OrdinalIgnoreCase) { "gzip" };
+
+            var selectedEncoding = StringSegment.Empty;
+            var selectedEncodingQuality = .0;
+
+            foreach (var encoding in encodings)
+            {
+                var encodingName = encoding.Value;
+                var quality = encoding.Quality.GetValueOrDefault(1);
+
+                if (quality < double.Epsilon)
+                {
+                    continue;
+                }
+
+                if (quality <= selectedEncodingQuality)
+                {
+                    continue;
+                }
+
+                if (supportedEncodings.Contains(encodingName))
+                {
+                    selectedEncoding = encodingName;
+                    selectedEncodingQuality = quality;
+                }
+
+                if (StringSegment.Equals("*", encodingName, StringComparison.Ordinal))
+                {
+                    selectedEncoding = supportedEncodings.First();
```

Comment:
This makes me wonder if `supportedEncodings` should really be a `HashSet`. If we're picking "first" from an unordered set, that implies we really don't mind which one we get.

Perhaps we should have a `preferredEncodings` ordered list and use `preferredEncodings.First()` here, and from that also compute `supportedEncodings` by just combining all its values into a hashset.

Of course, it doesn't start to matter until there's more than one supported encoding. But you must already be trying to account for that otherwise you'd have just hardcoded the idea that there's only one encoding and it's gzip.

Reply:
In fact more broadly, I wonder if we should consider simpler logic for the negotiation here. Something like:

1. Does the client accept `br` at all? If so, use that. (Obviously, skip this step until we support `br`.)
2. Does the client accept `gz` at all? If so, use that.
3. Send without compression.

It doesn't seem like there are use cases for letting the client express more detailed preferences than that.

Reply:
I've added a list to track our preference when the client doesn't have any. We could simplify things a bit, but performing content negotiation here is not complicated and that way we behave more inline with the rest of asp.net core.